### PR TITLE
Plumbing simplified import/export mechanism and path metadata API in storm

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

**Review by Cisphyx — APPROVED** [review]

**Review by Cisphyx — DISMISSED** [review]

File: synapse/cortex.py [hunk 1]
```diff
@@ -3673,6 +3693,80 @@ async def callStorm(self, text, opts=None):
         view = self._viewFromOpts(opts)
         return await view.callStorm(text, opts=opts)
 
+    async def export(self, text, opts=None):
+        opts = self._initStormOpts(opts)
+        user = self._userFromOpts(opts)
+        view = self._viewFromOpts(opts)
+
```
Comment (vEpiphyte) [inline synapse/cortex.py:3700]: this should be promoted to a task as well

File: synapse/lib/snap.py [hunk 2]
```diff
@@ -123,13 +151,23 @@ async def iterStormPodes(self, text, opts=None, user=None):
 
         self.core._logStormQuery(text, user)
 
+        scrubber = None
+        # NOTE: This option is still experimental and subject to change.
+        if opts.get('scrub') is not None:
```
Comment (vEpiphyte) [inline synapse/lib/snap.py:156]: is this intended mainly for the export use case; or for any remote storm call?

File: synapse/lib/spooled.py [hunk 3]
```diff
@@ -96,3 +96,59 @@ def discard(self, valu):
             return
 
         self.realset.discard(valu)
+
+class Dict(Spooled):
```
Comment (vEpiphyte) [inline synapse/lib/spooled.py:100]: missing a ``__len__`` implementation

File: synapse/lib/spooled.py [hunk 4]
```diff
@@ -96,3 +96,59 @@ def discard(self, valu):
...
+    async def __anit__(self, dirn=None, size=10000):
+
+        await Spooled.__anit__(self, dirn=dirn, size=size)
+        self.realdict = {}
+        self.len = 0
+
+    async def set(self, key, val):
+
+        if self.fallback:
+            self.slab.put(s_msgpack.en(key), s_msgpack.en(val))
```
Comment (vEpiphyte) [inline synapse/lib/spooled.py:111]: len is not updated in the fallback case

File: synapse/cortex.py [hunk 5]
```diff
@@ -3673,6 +3693,80 @@ async def callStorm(self, text, opts=None):
...
+    async def feedFromAxon(self, sha256, opts=None):
+
+        opts = self._initStormOpts(opts)
+        user = self._userFromOpts(opts)
+        view = self._viewFromOpts(opts)
+
+        await self.boss.promote('feeddata', user=user, info={'name': 'syn.nodes', 'sha256': sha256})
+
+        # ensure that the user can make all node edits in the layer
+        user.confirm(('node',), gateiden=view.layers[0].iden)
```
Comment (vEpiphyte) [inline synapse/cortex.py:3734]: This is a different permission check than the other feed apis use.

File: synapse/lib/stormtypes.py [hunk 6]
```diff
@@ -1646,8 +1646,26 @@ def getObjLocals(self):
...
+
+        Returns:
+            int: The number of nodes loaded.
+        '''
+        sha256 = await tostr(sha256)
+        opts = {
+            'user': self.runt.snap.user.iden,
+            'view': self.runt.snap.view.iden,
+        }
+        return await self.runt.snap.core.feedFromAxon(sha256, opts=opts)
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:1667]: should this be a dyncall?

File: synapse/tests/test_lib_spooled.py [hunk 7]
```diff
@@ -61,3 +61,21 @@ async def test_spooled_set_dir(self):
...
                 self.true(os.path.isdir(sset.slab.path))
                 self.true(os.path.abspath(sset.slab.path).startswith(dirn))
+
+    async def test_spooled_dict(self):
+
+        async def runtest(x):
+            await x.set(10, 'hehe')
+            await x.set(20, 'haha')
+            await x.set(30, 'hoho')
+            self.eq('hehe', x.get(10))
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_spooled.py:71]: len tests

File: synapse/lib/snap.py (hunk 2 above)
Comment (invisig0th) [reply synapse/lib/snap.py:156]: Intended for potential use in any storm call.

File: synapse/lib/stormtypes.py (hunk 6 above)
Comment (invisig0th) [reply synapse/lib/stormtypes.py:1667]: No, because the API is exposed via telepath.  Calling it on a remote or local cortex is the same.

**Review by vEpiphyte — DISMISSED** [review]

**Review by vEpiphyte — APPROVED** [review]

**Pull request merged**